pkg/operator/encryption/prune_controller.go

```diff
@@ -0,0 +1,226 @@
+// them.  Keeping a small number of old keys around is meant to help facilitate
+// decryption of old backups (and general precaution).
+type pruneController struct {
+	operatorClient operatorv1helpers.StaticPodOperatorClient
+
+	queue         workqueue.RateLimitingInterface
+	eventRecorder events.Recorder
+
+	preRunCachesSynced []cache.InformerSynced
+
+	encryptedGRs map[schema.GroupResource]bool
+
+	targetNamespace          string
+	encryptionSecretSelector metav1.ListOptions
+
+	podClient    corev1client.PodsGetter
+	secretClient corev1client.SecretsGetter
+}
+
+func newPruneController(
+	targetNamespace string,
+	operatorClient operatorv1helpers.StaticPodOperatorClient,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	podClient corev1client.PodsGetter,
+	secretClient corev1client.SecretsGetter,
+	encryptionSecretSelector metav1.ListOptions,
+	eventRecorder events.Recorder,
+	encryptedGRs map[schema.GroupResource]bool,
+) *pruneController {
+	c := &pruneController{
+		operatorClient: operatorClient,
+
+		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "EncryptionPruneController"),
+		eventRecorder: eventRecorder.WithComponentSuffix("encryption-prune-controller"), // TODO unused
+
+		encryptedGRs:    encryptedGRs,
+		targetNamespace: targetNamespace,
+
+		encryptionSecretSelector: encryptionSecretSelector,
+		podClient:                podClient,
+		secretClient:             secretClient,
+	}
+
+	c.preRunCachesSynced = setUpAllEncryptionInformers(operatorClient, targetNamespace, kubeInformersForNamespaces, c.eventHandler())
+
+	return c
+}
+
+func (c *pruneController) sync() error {
+	if ready, err := shouldRunEncryptionController(c.operatorClient); err != nil || !ready {
+		return err // we will get re-kicked when the operator status updates
+	}
+
+	// TODO do we want to use this to control the number we keep around?
+	// operatorSpec.SucceededRevisionLimit
+
+	configError := c.deleteOldMigratedSecrets()
+
+	// update failing condition
+	cond := operatorv1.OperatorCondition{
+		Type:   "EncryptionPruneControllerDegraded",
+		Status: operatorv1.ConditionFalse,
+	}
+	if configError != nil {
+		cond.Status = operatorv1.ConditionTrue
+		cond.Reason = "Error"
+		cond.Message = configError.Error()
+	}
+	if _, _, updateError := operatorv1helpers.UpdateStaticPodStatus(c.operatorClient, operatorv1helpers.UpdateStaticPodConditionFn(cond)); updateError != nil {
+		return updateError
+	}
+
+	return configError
+}
+
+func (c *pruneController) deleteOldMigratedSecrets() error {
+	_, encryptionState, _, err := getEncryptionConfigAndState(c.podClient, c.secretClient, c.targetNamespace, c.encryptionSecretSelector, c.encryptedGRs)
+	if err != nil {
+		return err
+	}
+	// do not try to delete anything during a transition state
+	if len(encryptionState) == 0 {
+		c.queue.AddAfter(pruneWorkKey, 2*time.Minute)
```

Comment:
does it make sense? If `encryptionState == 0` there are no secrets and this controller will be notified anyway as soon as a secret will be added.

Reply:
also the case of `APIServerRevisionNotConverged`. But we watch the pods and should be notified through them. I think we can safely remove this.